timer: add support for sigev_thread flag in zephyr timer posix api

arun-mani-tech opened the pull request:
add support for the sigev_thread flag in zephyr timer posix api, following the behavior described in the linux man page. with this enhancement, a single thread is created to receive all notifications from timers

cfriedt commented:
This is probably the specification we should refer to in this pr

https://pubs.opengroup.org/onlinepubs/9699919799/functions/timer_create.html

cfriedt commented on lib/posix/timer.c:
```diff
@@ -221,6 +280,15 @@ int timer_delete(timer_t timerid)
 	}
 
+	if (timer->sigev_notify == SIGEV_THREAD) {
+		pthread_cancel(timer->thread);
```
The spec explicitly says that the threads should be created detached. Therefore we should not call join.

cfriedt commented on lib/posix/timer.c:
```diff
@@ -75,6 +107,33 @@ int timer_create(clockid_t clockid, struct sigevent *evp, timer_t *timerid)
 	if (evp->sigev_notify == SIGEV_NONE) {
 		k_timer_init(&timer->ztimer, NULL, NULL);
+	} else if (evp->sigev_notify == SIGEV_THREAD) {
+		int ret;
```
There is no need to manually allocate a thread stack in this function.

Thread stack allocation is automatic with `pthread_create(&tid, NULL, func, arg)`. I.e. with a `NULL` `pthread_attr_t`.

cfriedt replied:
The spec says that users may pass in a `pthread_attr_t *` may be passed in via `sigev_notify_attributes` and if it is non-`NULL` then we should use the stack provided from those attributes.

cfriedt commented on lib/posix/timer.c:
```diff
@@ -16,7 +17,12 @@ static void zephyr_timer_wrapper(struct k_timer *ztimer);
+	struct k_sem sem_cond;
+	pthread_t thread;
+	pthread_attr_t attr;
+	struct z_thread_stack_element *stack;
```
This is unnecessary, since it is already contained within the `pthread_attr_t`

cfriedt commented on lib/posix/timer.c:
```diff
@@ -16,7 +17,12 @@ static void zephyr_timer_wrapper(struct k_timer *ztimer);
 	void (*sigev_notify_function)(sigval val);
+	struct k_sem sem_cond;
+	pthread_t thread;
+	pthread_attr_t attr;
```
This is unecessary to keep, and can be discarded after `pthread_create()` is called

cfriedt reviewed (changes requested):
There seems to be some misaligned from the spec. That needs to be addressed before it is considered.

Some other issues that seem unnecessary (unnecessary storage of things in a structure, unnecessary allocation, etc).

Please also create additional tests for the testsuite as a separate second commit to demonstrate and verify functionality.

cfriedt commented:
@arun-mani-tech - can you please rebase on latest `main` and resolve conflicts?

The tests that are failing are unrelated and passing on `main`

Commit:
timer: add support for sigev_thread flag in zephyr timer posix api

add support for the sigev_thread flag in zephyr timer posix api following
the behaviour described in the linux man page. with this enhancement,
a single thread is created to receive all notifications from timers

Signed-off-by: Arunmani Alagarsamy <arunmani@linumiz.com>

Commit:
tests: posix: common: timer: add test for SIGEV_THREAD

this flag enables the timer expiry handler runs on thread

Signed-off-by: Arunmani Alagarsamy <arunmani@linumiz.com>

cfriedt commented:
@arun-mani-tech - you may have better luck using pool allocated thread stacks rather than heap allocated thread stacks. Heap allocation can require more memory resources than what is available on certain platforms to satisfy alignments, MPU, etc.

See e.g. #60759 

```
CONFIG_POSIX_API=y
CONFIG_THREAD_STACK_INFO=y
CONFIG_DYNAMIC_THREAD=y
CONFIG_DYNAMIC_THREAD_POOL_SIZE=2
```

cfriedt commented on lib/posix/timer.c:
```diff
@@ -76,6 +106,19 @@ int timer_create(clockid_t clockid, struct sigevent *evp, timer_t *timerid)
+
+		timer->sigev_notify = SIGEV_THREAD;
+		k_sem_init(&timer->sem_cond, 0, 1);
+		ret = pthread_create(&timer->thread, NULL, zephyr_thread_wrapper, timer);
```
```suggestion
		ret = pthread_create(&timer->thread, evp->sigev_notify_attributes, zephyr_thread_wrapper, timer);
```

cfriedt commented:
@arun-mani-tech - you may have better luck using pool allocated thread stacks rather than heap allocated thread stacks. Heap allocation can require more memory resources than what is available on certain platforms to satisfy alignments, MPU, etc.

See e.g. #60759 

```
CONFIG_POSIX_API=y
CONFIG_THREAD_STACK_INFO=y
CONFIG_DYNAMIC_THREAD=y
CONFIG_DYNAMIC_THREAD_POOL_SIZE=2
```

cfriedt reviewed (changes requested):
Minor change still needed for the call to `pthread_create()` is to use the caller provided `sigev_notify_attributes`. If it is `NULL`, then a thread stack is automatically allocated. 

Otherwise, looks good. I would maybe use pooled / dynamic thread stacks since they are a bit easier on arches with complicated MPU alignments.

cfriedt commented:
One of the test failures leads me to believe that we should be filtering out `arch_posix` in a `testcase.yaml` file.

```
2023-09-08T14:56:32.5934159Z INFO    - -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
2023-09-08T14:56:32.5934859Z INFO    - The following issues were found (showing the top 10 items):
2023-09-08T14:56:32.5935559Z INFO    - 1) tests/posix/common/portability.posix.common.picolibc on nrf52_bsim failed (Failed)
2023-09-08T14:56:32.5935994Z INFO    - 
2023-09-08T14:56:32.5936484Z INFO    - To rerun the tests, call twister using the following commandline:
2023-09-08T14:56:32.5937047Z INFO    - west twister -p <PLATFORM> -s <TEST ID>, for example:
2023-09-08T14:56:32.5937443Z INFO    - 
2023-09-08T14:56:32.5937929Z INFO    - west twister -p nrf52_bsim -s tests/posix/common/portability.posix.common.picolibc
2023-09-08T14:56:32.5938363Z INFO    - or with west:

```

cfriedt commented:
it seems to fail on `mps2_an385`
```
In file included from /__w/zephyr/zephyr/lib/posix/timer.c:9:
2023-09-08T14:33:56.1660296Z /__w/zephyr/zephyr/include/zephyr/posix/pthread.h:445:18: error: unknown type name 'pthread_once_t'; did you mean 'pthread_cond_t'?
2023-09-08T14:33:56.1660840Z   445 | int pthread_once(pthread_once_t *once, void (*initFunc)(void));
2023-09-08T14:33:56.1661195Z       |                  ^~~~~~~~~~~~~~
```

cfriedt commented:
it seems to fail on `mps2_an385`
```
In file included from /__w/zephyr/zephyr/lib/posix/timer.c:9:
2023-09-08T14:33:56.1660296Z /__w/zephyr/zephyr/include/zephyr/posix/pthread.h:445:18: error: unknown type name 'pthread_once_t'; did you mean 'pthread_cond_t'?
2023-09-08T14:33:56.1660840Z   445 | int pthread_once(pthread_once_t *once, void (*initFunc)(void));
2023-09-08T14:33:56.1661195Z       |                  ^~~~~~~~~~~~~~
```

arun-mani-tech commented:
@cfriedt Changes have been made please revisit

arun-mani-tech commented:
@ycsin please review

arun-mani-tech commented:
@povergoing @jgl-meta Please review this

ycsin commented:
Not a fan of single line if-conditionals without curly braces, otherwise LGTM

povergoing reviewed (approved):
Overall LGTM, also not a fan of if-conditionals without curly braces
